### PR TITLE
guard save model hooks to only execute on main process

### PR DESCRIPTION
**Pull request by williamberman**
re internal conversation accelerate save/load model hooks execute on every process and have to be guarded if io requires guarding -- i.e. saving models

### CODE REVIEW

**Review by sayakpaul — APPROVED** [review]
Life is nothing but a sea of follow-up PRs and `main` processes.

File: examples/controlnet/train_controlnet.py [hunk 1]
```diff
@@ -785,16 +785,17 @@ def main(args):
...
+                    weights.pop()
+                    model = models[i]
 
-                sub_dir = "controlnet"
-                model.save_pretrained(os.path.join(output_dir, sub_dir))
+                    sub_dir = "controlnet"
+                    model.save_pretrained(os.path.join(output_dir, sub_dir))
 
-                i -= 1
+                    i -= 1
```
Comment (muellerzr) [inline examples/controlnet/train_controlnet.py:798]: Can't this be:

```python
if accelerator.is_main_process:
    for index in range(len(weights)):
        weights.pop()
        model = models[index]
        sub_dir = "controlnet"
        model.save_pretrained(os.path.join(output_dir, sub_dir)
````

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 2]
```diff
@@ -651,14 +651,15 @@ def compute_snr(timesteps):
     if version.parse(accelerate.__version__) >= version.parse("0.16.0"):
         # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
         def save_model_hook(models, weights, output_dir):
-            if args.use_ema:
-                ema_unet.save_pretrained(os.path.join(output_dir, "unet_ema"))
+            if accelerator.is_main_process:
+                if args.use_ema:
+                    ema_unet.save_pretrained(os.path.join(output_dir, "unet_ema"))
```
Comment (muellerzr) [inline examples/text_to_image/train_text_to_image_sdxl.py:656]:
```suggestion
            if accelerator.is_main_process and args.use_ema:
                ema_unet.save_pretrained(os.path.join(output_dir, "unet_ema"))
```

File: examples/unconditional_image_generation/train_unconditional.py [hunk 3]
```diff
@@ -309,14 +309,15 @@ def main(args):
     if version.parse(accelerate.__version__) >= version.parse("0.16.0"):
         # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
         def save_model_hook(models, weights, output_dir):
-            if args.use_ema:
-                ema_model.save_pretrained(os.path.join(output_dir, "unet_ema"))
+            if accelerator.is_main_process:
+                if args.use_ema:
```
Comment (muellerzr) [inline examples/unconditional_image_generation/train_unconditional.py:313]: Same here

**Review by muellerzr — COMMENTED** [review]
Thanks! overall this looks like the right approach/usage of `is_main_process`, left some comments in terms of code complexity and readability.

**Pull request merged**